respy/simulate.py

```diff
@@ -92,177 +136,196 @@ def simulate(params, base_draws_sim, base_draws_wage, state_space, options):
     """
     optim_paras, options = process_params_and_options(params, options)
 
+    # Solve the model.
     state_space.update_systematic_rewards(optim_paras)
-
     state_space = solve_with_backward_induction(state_space, optim_paras, options)
 
-    simulated_data = _simulate_data(
-        state_space, base_draws_sim, base_draws_wage, optim_paras, options
-    )
-
-    return simulated_data
+    # Adjust options and prepare data. If no data is passed or if only one observation
+    # for each individual is passed, perform n-step-ahead simulation. Else perform
+    # one-step-ahead simulation. Also, set a flag for the simulation method and adjust
+    # the number of periods.
+    if df is None:
```

Comment:
1. You are right, the functionality should be defined once. Another idea I had is to merge the two functions. Irrespective of whether the ``df`` has a certain characteristic, the characteristic is randomly sampled and then ``NaNs`` in the ``df`` are replaced with the sampled values.

   For n-step-ahead from initial conditions, all characteristics are added. For n-step-ahead from data or one-step-ahead normally only types are added. But, at the same time we would allow the model to integrate out missing characteristics in the first period.

    Since all characteristics are always sampled, the seeds between the operations are aligned. Thus, n-step-ahead with data and one-step-ahead would have the same types as n-step-ahead with initial conditions.

    What do you think about this?

2. Could you explain your second point a little bit more?

Reply:
I think this is a pretty good idea since it unifies the logic involved in this step!  The only potential barrier that I see are missing values in the input dfs. Would we also want to replace these or directly exclude such rows from the simulation?
  With respect to the second point I meant that if we could potentially have a model that contains two observables and a dataset that contains only one observable. Maybe people might want to simulate the second obs according to a prespecified distribution. It however seems to be a bit unnatural and that's why I am not  sure whether this should be explicitelly allowed for.

Reply:
So, your second point is actually also about imputing missing values. I do not think there is a big difference imputing only some values in some rows versus imputing a whole column. But you are right that we should be explicit about what happens.

I am not sure about deleting rows because if we can impute the missings, people should recognize this quickly.

What if there is another argument for the simulation interface which is ``impute_missings`` or ``sample_missings`` or something similar.

- By default it is ``None`` and if there missings in the passed data or columns missing, a warning is raised, but missings are imputed. I want to be user-friendly here, but maybe I am overshooting.
- Setting the value to ``True`` will silence the warning.
- Setting the value to ``False`` will raise an exception.

Since this critical to the interface, I am pinging @janosg. Short summary: The idea is to randomly sample every characteristic (experiences, lagged choices, obsevables) in the simulation regardless of simulation type (n-step, one-step). Then, missings in the existing or non-existing column are filled up with the sampled characteristics. Now, the question whether we want this kind of imputation by default or not.

Reply:
I like the interface and think it is important that there is a warning when values are imputed by default.

Reply:
I changed the interface in line with our comments, the only difference being that I did not implement ``replace_nans``. I raise a warning if the data in the first period contains NaNs, but it does not really make sense to implement another argument. The additional functionality is that a warning is silenced or an exception is raised. The user can do it herself when a warning is shown and we avoid the additional overhead in harmonizing arguments and checks.